### PR TITLE
[feat] 3/n pp

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds the doc for pp.

Fixes # (issue)

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/performance.rst [hunk 1]
```diff
@@ -131,4 +131,65 @@ To use Optimizer Sharded Training, refer to :ref:`model-parallelism`.
...
\ No newline at end of file
+Refer to the :ref:`distributed computing guide for more details <multi_gpu>`.
+
+
+Sequential Model Parallelism with Checkpointing to reduce peak memory
+---------------------------------------------------------------------
+Pipe Pipeline is a lightning integration of Pipeline Parallelism provided by Fairscale.
+Pipe combines pipeline parallelism with checkpointing to reduce peak memory required to train while minimizing device under-utilization.
+
+Find more explanation at https://arxiv.org/abs/1811.06965
```
Comment (SeanNaren) [inline docs/source/performance.rst:142]: As done for sharded, I'd add a cross reference to the docs rather than duplicating everything here.

Keep the memory benchmark, but cross reference everything else

File: docs/source/multi_gpu.rst [hunk 2]
```diff
@@ -678,6 +679,74 @@ Sharded Training can work across all DDP variants by adding the additional ``--p
...
+Sequential Model Parallelism splits a sequential module onto multiple GPUs, reducing peak GPU memory requirements substantially.
+We also provide auto-balancing techniques through FairScale, to find optimal balances for the model across GPUs.
+In addition, we use Gradient Checkpointing to reduce GPU memory requirements further, and micro-batches to minimizing device under-utilization automatically.
+
+Reference: https://arxiv.org/abs/1811.06965
+
+.. note:: DDPSequentialPlugin is currently supported only for Pytorch 1.6.
+
+To get started, install all extras using with ``pip install pytorch-lightning["extra"]``
+and Bolts with ``pip install pytorch-lightning-bolts``
```
Comment (carmocca) [inline docs/source/multi_gpu.rst:698]: question: Why is bolts needed? or which part?

Comment (SeanNaren) [reply docs/source/multi_gpu.rst:698]: Yeah exactly can we move this to pre-requisites for the example rather than required for the small example?

**Comment by SeanNaren** [comment]
Could we also add back the pip install command like sharded? (And update both to the latest zip file link!)

File: docs/source/multi_gpu.rst (hunk 2 above)
Comment (tchaton) [reply docs/source/multi_gpu.rst]: We are going to update this part. I moved Bolts to the example part where it is needed.

Comment (tchaton) [reply docs/source/multi_gpu.rst]: Can you review the doc again @carmocca @SeanNaren  ?

**Review by justusschock — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: docs/source/multi_gpu.rst [hunk 3]
```diff
@@ -678,6 +679,80 @@ Sharded Training can work across all DDP variants by adding the additional ``--p
...
+
+    # Split my module across 4 gpus, one layer each
+    model = MyModel()
+    plugin = DDPSequentialPlugin(balance=[1, 1, 1, 1])
+    trainer = Trainer(accelerator='ddp', gpus=4, plugins=[plugin])
+    trainer.fit(model)
+
+
+We provide a minimal example of Sequential Model Parallelism using a convolutional model training on cifar10, split onto GPUs `here <https://github.com/PyTorchLightning/pytorch-lightning/tree/master/pl_examples/basic_examples/conv_sequential_example.py>`_.
+To run the example, you will to install `Bolts <https://github.com/PyTorchLightning/pytorch-lightning-bolts>`_. Install with ``pip install pytorch-lightning-bolts``.
```
Comment (carmocca) [inline docs/source/multi_gpu.rst:726]:
```suggestion
To run the example, you need to install `Bolts <https://github.com/PyTorchLightning/pytorch-lightning-bolts>`_. Install with ``pip install pytorch-lightning-bolts``.
```

**Review by carmocca — APPROVED** [review]

File: docs/source/multi_gpu.rst [hunk 4]
```diff
@@ -678,6 +679,80 @@ Sharded Training can work across all DDP variants by adding the additional ``--p
...
+   * - Gpu 0
+     - 4436 MB
+     - 1554 MB
+   * - Gpu 1
+     - ~0
+     - 994 MB
+
+To run the example with Sequential Model Parallelism:
+
+.. code-block:: python
```
Comment (Borda) [inline docs/source/multi_gpu.rst:746]:
```suggestion
.. code-block:: bash
```

File: docs/source/multi_gpu.rst [hunk 5]
```diff
@@ -678,6 +679,80 @@ Sharded Training can work across all DDP variants by adding the additional ``--p
...
+
+    python pl_examples/basic_examples/conv_sequential_example.py --batch_size 1024 --gpus 2 --accelerator ddp --use_ddp_sequential
+
+To run the same example without Sequential Model Parallelism:
+
+.. code-block:: python
```
Comment (Borda) [inline docs/source/multi_gpu.rst:752]:
```suggestion
.. code-block:: bash
```

File: docs/source/multi_gpu.rst [hunk 6]
```diff
@@ -678,6 +679,80 @@ Sharded Training can work across all DDP variants by adding the additional ``--p
...
+     - 1554 MB
+   * - Gpu 1
+     - ~0
+     - 994 MB
+
+To run the example with Sequential Model Parallelism:
+
+.. code-block:: python
+
+    python pl_examples/basic_examples/conv_sequential_example.py --batch_size 1024 --gpus 2 --accelerator ddp --use_ddp_sequential
```
Comment (Borda) [inline docs/source/multi_gpu.rst:748]: if we have it in docs, we shall test this call

File: docs/source/training_tricks.rst [hunk 7]
```diff
@@ -123,3 +123,28 @@ The algorithm in short works by:
...
+    class MyModel(LightningModule):
+        def __init__(self):
+            ...
+            self.sequential_module = torch.nn.Sequential(my_layers)
+
+    # Split my module across 4 gpus, one layer each
+    model = MyModel()
+    plugin = DDPSequentialPlugin(balance=[1, 1, 1, 1])
+    trainer = Trainer(accelerator='ddp', gpus=4, plugins=[plugin])
+    trainer.fit(model)
```
Comment (Borda) [inline docs/source/training_tricks.rst:147]: it looks exactly the same, can we keep just one and refer to the second one...
lest work with maintaining and confusion if user reposts docs problem...

**Review by Borda — APPROVED** [review]

File: docs/source/multi_gpu.rst (hunk 6 above)
Comment (tchaton) [reply docs/source/multi_gpu.rst:748]: Yes, let's do this in follow up PR.
